feat(sidebar): add menu sidebar updating

Gelio opened the pull request:
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

A new feature - sidebar menu entries updating.

* **What is the current behavior?** (You can also link to an open issue here)

The sidebar menu is loaded at the beginning when _baSidebar.component_ is instantiated and remains unchanged since. It is also very hard to update it afterward.

* **What is the new behavior (if this is a feature change)?**

_baMenu.service_ is exposed to the application and via `updateMenuByRoutes`, which takes in the object of the type similar to _app.menu.ts_(essentially how the menu is structured), allows easy sidebar menu updating.

* **Other information**:

I believe this change does no harm and only provides benefits, as this feature was requested by a few people in issue #236.
If you have any questions, opinions or suggestions I would be happy to hear them.

To use this in a component/service of yours simply import `baMenu.service` from the _theme_ directory, inject it into the constructor of your component/service and use the `updateMenuByRoutes` method when the menu needs to be modified. You don't have to clone your menu, it will remain unchanged, opposed to the default behavior or _baMenu.service_ methods that mutate their inputs.

To test if this works you can modify the `constructor` of the _baMenu.service_:
``` typescript
  constructor(private _router:Router) {
    this.updateMenuByRoutes(<Routes>MENU);

    setInterval(() => {
      let menuPart = _.cloneDeep(MENU);
      menuPart[0].children = menuPart[0].children.filter(() => Math.random() > 0.5);
      console.log('Updating routes', menuPart);
      this.updateMenuByRoutes(<Routes>menuPart);
    }, 4000);
  }
```
Every 4 seconds the menu will be updated, items that should remain will be chosen randomly.

abedzantout commented:
Hello Gelio, just curious, why did you remove lodash import? is it imported somewhere else?

Gelio commented:
@abedzantout I think it's imported as a vendor dependency. Removing it caused no errors and moreover less code to maintain is overall for the better.

abedzantout commented:
@Gelio It caused an error from my side. The way i fixed it was adding declare var _: any; in custom_typings.d.ts

Gelio commented:
@abedzantout That's strange. I was able to successfully run that without any errors on the most recent master branch. Are you perhaps any commits behind master?

abedzantout commented:
@Gelio yes i noticed that is way i asked. The only thing differing are the packages that i am using, i am using lodash 4.17.2 and lodash types 4.14.38. Maybe this could be the reason.

nnixaa commented:
Hey @Gelio, this is great, many thanks!

While we are on it, probably it would make more sense to move the https://github.com/akveo/ng2-admin/pull/548/files#diff-49e9fb1e5d9b20de76e3c3e0cec8defeR15 part into the app component?
So that at the end the menu service becomes a global one and menu component won't depend on the app at all (will become an independent) and thus we will device the source of the menu items from the menu component.

What do you think?

Gelio commented:
@nnixaa 👍 Yes, definitely, that's a great suggestion! I knew it should be structured differently but I just didn't have any idea how to initialize the menu. I will change it later today and let you know when I'm finished

Gelio commented:
@nnixaa I've moved initializing the menu into _app.component.ts_. While doing so I found out that _baMenuService_ was in component's providers instead of module's ones, so it was impossible to change the menu outside of _baMenu.component_. I fixed that as well.

I'm looking forward to seeing your opinion on this.

nnixaa commented:
@Gelio probably let's move the menu service into the services folder https://github.com/akveo/ng2-admin/tree/master/src/app/theme/services? So that it becomes more of a global service, rather than a local component service.

Gelio commented:
Sure, I'll do it. Do you have any other suggestions?

19.12.2016 2:23 PM "Dmitry Nehaychik" <notifications@github.com> napisał(a):

> @Gelio <https://github.com/Gelio> probably let's move the menu service
> into the services folder https://github.com/akveo/ng2-
> admin/tree/master/src/app/theme/services? So that it becomes more of a
> global service, rather than a local component service.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/akveo/ng2-admin/pull/548#issuecomment-267962865>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA2SJ6J84dEM6Z-sHFdEV1hfO1HgB7_oks5rJoVXgaJpZM4LP44E>
> .
>

Gelio commented:
@nnixaa I've moved _baMenu.service_ to a separate directory in the _services_ directory, just as you asked. If there's anything more I can do just let me know

nnixaa commented:
@Gelio just merged this, many thanks again!

Gelio commented:
Not a problem, I'm glad I could help!

abedzantout commented:
It is worth noting that for those who are wondering how the test provided by @Gelio at first works after the additional changes he made. Modify the constructor in `app.component.ts` :

`
constructor(private _state: GlobalState,
	             private _imageLoader: BaImageLoaderService,
	             private _spinner: BaThemeSpinner,
	             private _config: BaThemeConfig,
	             private _menuService: BaMenuService,
	             private viewContainerRef: ViewContainerRef ) {

    this._menuService.updateMenuByRoutes(<Routes>MENU);

     setInterval(() => {
			let menuPart           = _.cloneDeep(MENU);
			menuPart[ 0 ].children = menuPart[ 0 ].children.filter(() => Math.random() > 0.5);
			console.log('Updating routes', menuPart);
			this._menuService.updateMenuByRoutes(<Routes>menuPart);
	}, 4000);`

Happy coding!

Lynx12 commented:
Could you review again this fix, i have tested the last changes but they are not working. The menu is not expanded when i write the URL and the component "baContentTop" is not loading the variable "activePageTitle".

raposok commented:
@Gelio Thank you for the sidebar update feature.
I'm new to Angular2 and Typescript and I'm having trouble implementing the update feature. I'm trying to pull the navigation in via URL using a service but haven't been able to figure out how to use the updateMenuByRoutes to modify the navigation. Any advice on what to do next is greatly appreciated or if anyone has an example that I could learn from.

_Modified app/pages/pages.menu.ts:_
```export const PAGES_MENU = [
  {
    path: 'pages',
    children: [
      {
        path: 'summary', 
        data: { 
          menu: {
            title: 'Summary', 
            icon: 'ion-android-home',
            selected: false,
            expanded: false,
            order: 100
          }
        }
      },
      {
        path: 'dashboard',
        data: {
          menu: {
            title: 'Project',
            icon: 'ion-android-radio-button-on',
            selected: false,
            expanded: true,
            order: 0
          }
        }
        
      }
    ]
  }
];```

**Navigation Service app/theme/services/nav/nav.service.ts:**
```import {Injectable} from '@angular/core';
import {Router, Routes} from '@angular/router';
import { Http, Response, Headers, RequestOptions } from '@angular/http';

import {Observable} from 'rxjs/Rx';

import {BaMenuService} from '../baMenu/baMenu.service';
import {BaMenu} from '../../components/baMenu/baMenu.component';
import { MENU } from '../../app.menu';

@Injectable()
export class NavService {

	constructor (
		private _http: Http,
		private _menuService: BaMenuService,
		private _router: Router
	)
		{
			this._menuService.updateMenuByRoutes(<Routes>MENU);
		}


	getNavs() {
		return this._http.get(this.navsUrl)
			.map((res:Response) => <Routes[]>res.json()
			.subscribe(onNext, error => console.log("An error occurred when requesting nav.json.", error));
	}
}```

**Navigation file nav.json:**
```[
  {
    "path": "pages",
    "children": [
      {
        "path": "dashboard",
        "data": {
          "menu": {
            "title": "Project1",
            "icon": "ion-android-radio-button-on",
            "selected": false,
            "expanded": true,
            "order": 0
          }
        }
      },
      {
        "path": "dashboard",
        "data": {
          "menu": {
            "title": "Project2",
            "icon": "ion-android-radio-button-on",
            "selected": false,
            "expanded": true,
            "order": 0
          }
        }
      }
    ]
  }
]```

Gelio commented:
@raposok Unfortunately, I haven't used this feature much. I recommend you to learn basics of Javascript and Typescript first before trying out Angular.

From what I can tell, your _NavService_ contains the following mistakes:
1. `getNavs` returns a subscription, not an Observable, which I guess is the intended behavior. Consider replacing the `subscribe` function with the [do](https://github.com/Reactive-Extensions/RxJS/blob/master/doc/api/core/operators/do.md) operator.
2. In `getNavs` you're missing a closing bracket on the line with the _map_ operator.
3. In `getNavs` the `onNext` function is not defined, which surely produces compilation and runtime errors. The `navsUrl` property of the service is not defined as well. You should consider setting it first.

Moreover your import of a lot of variables is redundant (namely _Headers_, _RequestOptions_, _Observable_, _BaMenu_).
